### DISCUSSION TITLE
Allow use of symlinks instead of junctions on Windows

### DISCUSSION THREAD

**Discussion by stkb** [pnpm/pnpm#3516] [Ideas]
Junctions can only have absolute paths, which means links to a project's virtual store break if you move the project folder (#895).

Symlinks have been creatable under normal permissions in Windows 10, in developer mode, for a couple of years now. 
Could there be a config option to force the use of (relative) symlinks rather than junctions, thereby solving that problem?

An alternative would be: the first time pnpm creates a link on Windows, it tries to create a symlink, and if that fails with an `EPERM` error, fall back to junctions instead.

I'd be willing to help in implementing this.

**Comment by zkochan** [comment]
I think a config would be fine.

**Reply by stkb** [reply]
Ok cool 👍, what's the next step then? Should I move this to an issue? Can I start hacking on a PR right away?

**Reply by zkochan** [reply]
I am not sure. Currently symlinking is done by this library: https://github.com/zkochan/symlink-dir

so it will have to accept an option

**Reply by stkb** [reply]
Yeah nothing is as simple as it first seems... I saw that after looking through the code. So these are the possible changes to `symlink-dir` that come to mind.

These require an API change / config to be passed in from `pnpm`:
1. Create an optional extra parameter for `symlinkDir()` that's passed in every time. Likely to be forgotten sometimes?
2. Make the module a "constructor" function instead that takes the option<sup>1</sup>. Maybe slightly cleaner and less prone to being forgotten than (1).

These require no config or extra input from pnpm / no api change to `symlink-dir` / just work:

3. Each time `symlinkDir()` is called, try creating a symlink, and if that fails with an EPERM error, try a junction instead. The "purest" but will have a performance penalty, though can't be sure how great without testing.
4. The same as (3) but after the first invocation, if creating a symlink fails, store the junction preference as state within the module and use junctions from then on. Only a problem if you don't like the idea of the module saving state?
5. Detect whether the machine is in developer mode by reading the appropriate registry key<sup>2</sup> and if so, create symlinks, otherwise create junctions. It should work, but if for some reason it doesn't, it could be combined with (3 [or 4]). Then only machines in a "weird state" will suffer from the performance penalty.

Further than that you have to decide what to do if you're creating a symlink but a junction already exists to the same target etc. but I think it's quite possible.

-----
1. Backward compatibility can still be preserved.
2. `HKLM\SOFTWARE\Microsoft\Windows\CurrentVersion\AppModelUnlock\AllowDevelopmentWithoutDevLicense` should have the numeric value `1`

**Comment by tats-u** [comment]
pnpm v11 uses genuine symlinks as long as you have turned on Developer Mode. It has not been mentioned in the release note of pnpm v11.0.0 yet though.
